File: cpp/src/virtual_machine.cpp
#include "virtual_machine.hpp"

#include <cmath>
#include <iostream>

#include "./values/value_property_access.hpp"
#include "./values/object_value.hpp"
#include "./standard_library/standard_array_library.hpp"
#include "./utils.hpp"
#include "./errors/virtual_machine_error.hpp"
#include "./errors/error_common.hpp"

namespace lysithea_vm
{
    std::shared_ptr<const array_value> virtual_machine::empty_args(std::make_shared<const array_value>(true));

    virtual_machine::virtual_machine(int stack_size) :
        stack(stack_size), stack_trace(stack_size), program_counter(0), running(false), paused(false),
        global_scope(std::make_shared<scope>())
    {
        current_scope = global_scope;
    }

    void virtual_machine::reset()
    {
        program_counter = 0;
        global_scope = std::make_shared<scope>();
        current_scope = global_scope;
        stack.clear();
        stack_trace.clear();
        running = false;
        paused = false;
    }

    void virtual_machine::change_to_script(std::shared_ptr<script> script)
    {
        program_counter = 0;
        stack.clear();
        stack_trace.clear();

        builtin_scope = script->builtin_scope;
        current_code = script->code;
    }

    void virtual_machine::execute(std::shared_ptr<script> script)
    {
        change_to_script(script);

        running = true;
        paused = false;

        while (running && !paused)
        {
            step();
        }
    }

    void virtual_machine::step()
    {
        if (program_counter >= current_code->code.size())
        {
            if (!try_return())
            {
                running = false;
            }
            return;
        }

        const auto &code_line = current_code->code[program_counter++];

        switch (code_line.op)
        {
            default:
            {
                throw virtual_machine_error(create_stack_trace(), "Unknown operator");
            }
            case vm_operator::push:
            {
                if (!code_line.value.is_undefined())
                {
                    stack.push(code_line.value);
                }
                else
                {
                    throw virtual_machine_error(create_stack_trace(), "Push needs an input");
                }
                break;
            }
            case vm_operator::to_argument:
            {
                auto top = get_operator_arg<array_value>(code_line);
                if (!top)
                {
                    throw virtual_machine_error(create_stack_trace(), std::string("Unable to convert input to argument: ") + top->to_string());
                }

                push_stack(std::make_shared<array_value>(top->data, true));
                break;
            }
            case vm_operator::get:
            {
                auto key = get_operator_arg(code_line);
                auto is_string = key.get_complex<string_value>();
                if (!is_string)
                {
                    throw virtual_machine_error(create_stack_trace(), std::string("Unable to get value, input needs to be a string: ") + key.to_string());
                }

                value found_value;
                if (current_scope->try_get_key(is_string->data, found_value) ||
                    (builtin_scope && builtin_scope->try_get_key(is_string->data, found_value)))
                {
                    push_stack(found_value);
                }
                else
                {
                    throw virtual_machine_error(create_stack_trace(), std::string("Unable to find value to get: ") + key.to_string());
                }
                break;
            }
            case vm_operator::get_property:
            {
                auto key = get_operator_arg<array_value>(code_line);
                if (!key)
                {
                    throw virtual_machine_error(create_stack_trace(), std::string("Unable to get property, input needs to be an array: ") + key->to_string());
                }

                auto top = pop_stack();
                value found;
                if (try_get_property(top, *key, found))
                {
                    push_stack(found);
                }
                else
                {
                    throw virtual_machine_error(create_stack_trace(), std::string("Unable to get property: ") + key->to_string());
                }

                break;
            }
            case vm_operator::define:
            {
                auto key = get_operator_arg(code_line);
                auto value = pop_stack();
                current_scope->try_define(key.to_string(), value);
                break;
            }
            case vm_operator::set:
            {
                auto key = get_operator_arg(code_line);
                auto value = pop_stack();
                if (!current_scope->try_set(key.to_string(), value))
                {
                    throw virtual_machine_error(create_stack_trace(), "Unable to set variable that has not been defined: " + key.to_string());
                }
                break;
            }
            case vm_operator::jump_false:
            {
                const auto label = get_operator_arg(code_line);
                auto top = pop_stack();
                if (top.is_false())
                {
                    jump(label.to_string());
                }
                break;
            }
            case vm_operator::jump_true:
            {
                const auto label = get_operator_arg(code_line);
                auto top = pop_stack();
                if (top.is_true())
                {
                    jump(label.to_string());
                }
                break;
            }
            case vm_operator::jump:
            {
                const auto label = get_operator_arg(code_line);
                jump(label.to_string());
                break;
            }
            case vm_operator::call_return:
            {
                call_return();
                break;
            }
            case vm_operator::call:
            {
                if (!code_line.value.is_number())
                {
                    throw virtual_machine_error(create_stack_trace(), "Call needs a num args code line input");
                }

                auto top = pop_stack();
                if (top.is_function())
                {
                    call_function(*top.get_complex(), code_line.value.get_int(), true);
                }
                else
                {
                    throw virtual_machine_error(create_stack_trace(), "Call needs a function to run");
                }
                break;
            }
            case vm_operator::call_direct:
            {
                auto error = false;
                if (!code_line.value.is_array())
                {
                    throw virtual_machine_error(create_stack_trace(), "Call direct needs an array input");
                }

                auto array_input = code_line.value.get_complex<const array_value>();
                if (array_input->data.size() != 2 ||
                    !array_input->data[0].is_function())
                {
                    throw virtual_machine_error(create_stack_trace(), "Call direct needs two inputs of func and number");
                }

                auto num_args = array_input->data[1];
                if (!num_args.is_number())
                {
                    throw virtual_machine_error(create_stack_trace(), "Call direct needs two inputs of func and number");
                }

                call_function(*array_input->data[0].get_complex(), num_args.get_int(), true);
                break;
            }

            // Misc Operator
            case vm_operator::string_concat:
            {
                if (!code_line.value.is_number())
                {
                    throw virtual_machine_error(create_stack_trace(), "StringConcat operator needs the number of args to concat");
                }

                auto args = get_args(code_line.value.get_int());
                std::stringstream ss;
                for (auto iter : args->data)
                {
                    ss << iter.to_string();
                }
                push_stack(ss.str());
                break;
            }

            // Math Operators
            case vm_operator::add:
            {
                push_stack(get_operator_num(code_line) + pop_stack_number());
                break;
            }

            case vm_operator::sub:
            {
                auto right = get_operator_num(code_line);
                auto left = pop_stack_number();
                push_stack(left - right);
                break;
            }

            case vm_operator::unary_negative:
            {
                push_stack(-pop_stack_number());
                break;
            }

            case vm_operator::multiply:
            {
                push_stack(get_operator_num(code_line) * pop_stack_number());
                break;
            }

            case vm_operator::divide:
            {
                auto right = get_operator_num(code_line);
                auto left = pop_stack_number();
                push_stack(left / right);
                break;
            }

            case vm_operator::inc:
            {
                if (!code_line.value.is_complex())
                {
                    throw virtual_machine_error(create_stack_trace(), "Inc operator needs code line variable");
                }

                auto key = code_line.value.to_string();
                double found_value;
                if (!current_scope->try_get_number(key, found_value))
                {
                    throw virtual_machine_error(create_stack_trace(), "Inc operator could not find variable or was not a number");
                }
                current_scope->try_set(key, value(found_value + 1.0));
                break;
            }

            case vm_operator::dec:
            {
                if (!code_line.value.is_complex())
                {
                    throw virtual_machine_error(create_stack_trace(), "Dec operator needs code line variable");
                }

                auto key = code_line.value.to_string();
                double found_value;
                if (!current_scope->try_get_number(key, found_value))
                {
                    throw virtual_machine_error(create_stack_trace(), "Dec operator could not find variable or was not a number");
                }
                current_scope->try_set(key, value(found_value - 1.0));
                break;
            }

            // Comparison Operators
            case vm_operator::less_than:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) < 0);
                break;
            }
            case vm_operator::less_than_equals:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) <= 0);
                break;
            }
            case vm_operator::equals:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) == 0);
                break;
            }
            case vm_operator::not_equals:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) != 0);
                break;
            }
            case vm_operator::greater_than:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) > 0);
                break;
            }
            case vm_operator::greater_than_equals:
            {
                auto right = get_operator_arg(code_line);
                auto left = pop_stack();
                push_stack(left.compare_to(right) >= 0);
                break;
            }

            // Boolean Operators
            case vm_operator::op_and:
            {
                push_stack(get_operator_bool(code_line) && pop_stack_bool());
                break;
            }
            case vm_operator::op_or:
            {
                push_stack(get_operator_bool(code_line) || pop_stack_bool());
                break;
            }
            case vm_operator::op_not:
            {
                push_stack(!pop_stack_bool());
                break;
            }

            // Value Create
            case vm_operator::make_array:
            {
                if (!code_line.value.is_number())
                {
                    throw virtual_machine_error(create_stack_trace(), "MakeArray operator needs the number of args to pop");
                }

                auto args = get_args(code_line.value.get_int());
                push_stack(array_value::make_value(args->data));
                break;
            }
            case vm_operator::make_object:
            {
                if (!code_line.value.is_number())
                {
                    throw virtual_machine_error(create_stack_trace(), "MakeObject operator needs the number of args to pop");
                }

                auto args = get_args(code_line.value.get_int());
                push_stack(object_value::join(*args));
                break;
            }
        }
    }

    std::shared_ptr<const array_value> virtual_machine::get_args(int num_args)
    {
        if (num_args == 0)
        {
            array_vector empty;
            return empty_args;
        }

        auto has_arguments = false;
        array_vector temp(num_args);
        for (auto i = 0; i < num_args; i++)
        {
            auto value = pop_stack();
            auto is_arg = value.get_complex<const array_value>();
            if (is_arg && is_arg->is_arguments_value)
            {
                has_arguments = true;
            }
            temp[num_args - i - 1] = value;
        }

        if (has_arguments)
        {
            array_vector combined;
            for (const auto &iter : temp)
            {
                auto is_arg = iter.get_complex<const array_value>();
                if (is_arg && is_arg->is_arguments_value)
                {
                    for (const auto &arg_iter : is_arg->data)
                    {
                        combined.push_back(arg_iter);
                    }
                }
                else
                {
                    combined.push_back(iter);
                }
            }

            return std::make_shared<const array_value>(combined, true);
        }

        return std::make_shared<const array_value>(temp, true);
    }

    void virtual_machine::jump(const std::string &label)
    {
        auto find = current_code->labels.find(label);
        if (find == current_code->labels.cend())
        {
            throw virtual_machine_error(create_stack_trace(), std::string("Unable to jump to label: ") + label);
        }

        program_counter = find->second;
    }

    void virtual_machine::call_function(const complex_value &value, int num_args, bool push_to_stack_trace)
    {
        if (!value.is_function())
        {
            throw virtual_machine_error(create_stack_trace(), std::string("Unable to invoke non function value") + value.to_string());
        }
        auto args = get_args(num_args);
        value.invoke(*this, args, push_to_stack_trace);
    }

    void virtual_machine::execute_function(std::shared_ptr<function> code, std::shared_ptr<const array_value> args, bool push_to_stack_trace)
    {
        if (push_to_stack_trace)
        {
            push_stack_trace(scope_frame(program_counter, current_code, current_scope));
        }

        current_code = code;
        current_scope = std::make_shared<scope>(current_scope);
        program_counter = 0;

        auto num_called_args = std::min(args->data.size(), code->parameters.size());
        auto i = 0;
        for (; i < num_called_args; i++)
        {
            const auto &arg_name = code->parameters[i];
            auto is_unpack = starts_with_unpack(arg_name);
            if (is_unpack)
            {
                current_scope->try_define(arg_name.substr(3), standard_array_library::sublist(args->data, i, -1));
                i++;
                break;
            }
            current_scope->try_define(arg_name, args->data[i]);
        }

        if (i < code->parameters.size())
        {
            const auto &arg_name = code->parameters[i];
            auto is_unpack = starts_with_unpack(arg_name);
            if (is_unpack)
            {
                current_scope->try_define(arg_name.substr(3), array_value::empty);
            }
            else
            {
                throw virtual_machine_error(create_stack_trace(), "Function called without enough arguments");
            }
        }
    }

    bool virtual_machine::try_return()
    {
        scope_frame top;
        if (!stack_trace.pop(top))
        {
            return false;
        }

        current_code = top.code;
        current_scope = top.frame_scope;
        program_counter = top.line_counter;
        return true;
    }

    void virtual_machine::call_return()
    {
        if (!try_return())
        {
            throw virtual_machine_error(create_stack_trace(), "Unable to return, call stack empty");
        }
    }

    void virtual_machine::print_stack_debug()
    {
        const auto &data = stack.stack_data();
        std::cout << "Stack size: " << data.size() << "\n";
        for (const auto &iter : data)
        {
            std::cout << "- " << iter.to_string() << "\n";
        }
    }

    void virtual_machine::print_stack_trace_debug()
    {
        const auto &data = create_stack_trace();
        for (const auto &iter : data)
        {
            std::cout << iter << std::endl;
        }
    }

    std::vector<std::string> virtual_machine::create_stack_trace()
    {
        std::vector<std::string> result;

        result.emplace_back(debug_scope_line(*current_code, program_counter - 1));
        const auto &stack_data = stack_trace.stack_data();
        for (auto i = stack_trace.stack_size() - 1; i >= 0; i--)
        {
            const auto &stack_frame = stack_data[i];
            result.emplace_back(debug_scope_line(*stack_frame.code, stack_frame.line_counter - 1));
        }

        return result;
    }

    std::string virtual_machine::debug_scope_line(const function &func, int line)
    {
        std::stringstream ss;
        ss << "  at [" << func.name << "] in " << func.symbols->source_name;
        if (line >= func.code.size())
        {
            ss << " end of code";
            return ss.str();
        }
        else if (line < 0)
        {
            ss << " before start of code";
            return ss.str();
        }

        code_location location;
        func.symbols->try_get_location(line, location);

        ss << create_error_log_at(func.symbols->source_name, location, *func.symbols->full_text);
        return ss.str();
    }
} // namespace lysithea_vm

File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/VirtualMachine.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

#nullable enable

namespace SimpleStackVM
{
    public class VirtualMachine
    {

        #region Fields
        private readonly FixedStack<IValue> stack;
        private readonly FixedStack<ScopeFrame> stackTrace;

        public IReadOnlyScope? BuiltinScope = null;
        public Scope GlobalScope { get; private set; }
        public Scope CurrentScope { get; private set; }
        private int lineCounter = 0;

        public bool Running;
        public bool Paused;
        public Function CurrentCode = Function.Empty;

        public IReadOnlyFixedStack<IValue> Stack => this.stack;
        public IReadOnlyFixedStack<ScopeFrame> StackTrace => this.stackTrace;
        #endregion

        #region Constructor
        public VirtualMachine(int stackSize)
        {
            this.GlobalScope = new Scope();
            this.CurrentScope = this.GlobalScope;
            this.stack = new FixedStack<IValue>(stackSize);
            this.stackTrace = new FixedStack<ScopeFrame>(stackSize);
        }
        #endregion

        #region Methods
        public void Reset()
        {
            this.GlobalScope = new Scope();
            this.CurrentScope = this.GlobalScope;
            this.lineCounter = 0;
            this.stack.Clear();
            this.stackTrace.Clear();
            this.Running = false;
            this.Paused = false;
        }

        public void ChangeToScript(Script script)
        {
            this.lineCounter = 0;
            this.stack.Clear();
            this.stackTrace.Clear();

            this.BuiltinScope = script.BuiltinScope;
            this.CurrentCode = script.Code;
        }

        public void Execute(Script script)
        {
            this.ChangeToScript(script);

            this.Running = true;
            this.Paused = false;
            while (this.Running && !this.Paused)
            {
                this.Step();
            }
        }

        public void Step()
        {
            if (this.lineCounter >= this.CurrentCode.Code.Count)
            {
                if (!this.TryReturn())
                {
                    this.Running = false;
                }
                return;
            }

            var codeLine = this.CurrentCode.Code[this.lineCounter++];

            // this.PrintStackDebug();

            switch (codeLine.Operator)
            {
                default:
                    {
                        throw new UnknownOperatorException(this.CreateStackTrace(), $"Unknown operator: {codeLine.Operator}");
                    }
                case Operator.Push:
                    {
                        if (codeLine.Input != null)
                        {
                            this.PushStack(codeLine.Input);
                        }
                        else
                        {
                            throw new StackException(this.CreateStackTrace(), "Push needs an input");
                        }

                        break;
                    }
                case Operator.Get:
                    {
                        var key = codeLine.Input ?? this.PopStack();
                        if (!(key is StringValue stringInput))
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Unable to get variable, input needs to be a string: {key.ToString()}");
                        }

                        var keyString = key.ToString();
                        if (this.CurrentScope.TryGetKey(keyString, out var foundValue) ||
                            (this.BuiltinScope != null && this.BuiltinScope.TryGetKey(keyString, out foundValue)))
                        {
                            this.PushStack(foundValue);
                        }
                        else
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Unable to get variable: {key.ToString()}");
                        }
                        break;
                    }
                case Operator.GetProperty:
                    {
                        var key = codeLine.Input ?? this.PopStack();
                        if (!(key is ArrayValue arrayInput))
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Unable to get property, input needs to be an array: {key.ToString()}");
                        }

                        var top = this.PopStack();
                        if (ValuePropertyAccess.TryGetProperty(top, arrayInput, out var found))
                        {
                            this.PushStack(found);
                        }
                        else
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Unable to get property: {key.ToString()}");
                        }
                        break;
                    }
                case Operator.Define:
                    {
                        var key = codeLine.Input ?? this.PopStack();
                        var value = this.PopStack();
                        this.CurrentScope.Define(key.ToString(), value);
                        break;
                    }
                case Operator.Set:
                    {
                        var key = codeLine.Input ?? this.PopStack();
                        var value = this.PopStack();
                        if (!this.CurrentScope.TrySet(key.ToString(), value))
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Unable to set variable that has not been defined: {key.ToString()} = {value.ToString()}");
                        }
                        break;
                    }
                case Operator.JumpFalse:
                    {
                        var label = codeLine.Input ?? this.PopStack();

                        var top = this.PopStack();
                        if (top.CompareTo(BoolValue.False) == 0)
                        {
                            this.Jump(label.ToString());
                        }
                        break;
                    }
                case Operator.JumpTrue:
                    {
                        var label = codeLine.Input ?? this.PopStack();
                        var top = this.PopStack();
                        if (top.CompareTo(BoolValue.True) == 0)
                        {
                            this.Jump(label.ToString());
                        }
                        break;
                    }
                case Operator.Jump:
                    {
                        var label = codeLine.Input ?? this.PopStack();
                        this.Jump(label.ToString());
                        break;
                    }
                case Operator.Return:
                    {
                        this.Return();
                        break;
                    }
                case Operator.Call:
                    {
                        if (codeLine.Input == null || !(codeLine.Input is NumberValue numArgs))
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Call needs a num args code line input");
                        }

                        var top = this.PopStack();
                        if (top is IFunctionValue procTop)
                        {
                            this.CallFunction(procTop, numArgs.IntValue, true);
                        }
                        else
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Call needs a function to run: {top.ToString()}");
                        }
                        break;
                    }
                case Operator.CallDirect:
                    {
                        if (codeLine.Input == null || !(codeLine.Input is ArrayValue arrayInput) ||
                           !(arrayInput[0] is IFunctionValue procTop) || !(arrayInput[1] is NumberValue numArgs))
                        {
                            throw new OperatorException(this.CreateStackTrace(), $"Call direct needs an array of the function and num args code line input");
                        }

                        this.CallFunction(procTop, numArgs.IntValue, true);
                        break;
                    }
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void Jump(string label)
        {
            if (this.CurrentCode.Labels.TryGetValue(label, out var line))
            {
                this.lineCounter = line;
            }
            else
            {
                throw new OperatorException(this.CreateStackTrace(), $"Unable to jump to label: {label}");
            }
        }

        #region Function Methods
        public ArrayValue GetArgs(int numArgs)
        {
            var args = ArrayValue.Empty;
            if (numArgs > 0)
            {
                var temp = new IValue[numArgs];
                for (var i = 0; i < numArgs; i++)
                {
                    temp[numArgs - i - 1] = this.PopStack();
                }
                args = new ArrayValue(temp);
            }
            return args;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void CallFunction(IFunctionValue value, int numArgs, bool pushToStackTrace)
        {
            value.Invoke(this, numArgs, pushToStackTrace);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void ExecuteFunction(Function function, int numArgs = -1, bool pushToStackTrace = false)
        {
            if (pushToStackTrace)
            {
                this.PushToStackTrace(new ScopeFrame(this.CurrentCode, this.CurrentScope, this.lineCounter));
            }

            this.CurrentCode = function;
            this.CurrentScope = new Scope(this.CurrentScope);
            this.lineCounter = 0;

            var args = this.GetArgs(numArgs >= 0 ? Math.Min(numArgs, function.Parameters.Count) : function.Parameters.Count);
            for (var i = 0; i < args.Length; i++)
            {
                var argName = function.Parameters[i];
                this.CurrentScope.Define(argName, args[i]);
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void PushToStackTrace(ScopeFrame scopeFrame)
        {
            if (!this.stackTrace.TryPush(scopeFrame))
            {
                throw new StackException(this.CreateStackTrace(), "Unable to call, call stack full");
            }
        }

        public bool TryReturn()
        {
            if (!this.stackTrace.TryPop(out var scopeFrame))
            {
                return false;
            }

            this.CurrentCode = scopeFrame.Function;
            this.CurrentScope = scopeFrame.Scope;
            this.lineCounter = scopeFrame.LineCounter;
            return true;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void Return()
        {
            if (!this.TryReturn())
            {
                throw new StackException(this.CreateStackTrace(), "Unable to return, call stack empty");
            }
        }

        #endregion

        #region Stack Methods
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public IValue PopStack()
        {
            if (!this.stack.TryPop(out var obj))
            {
                throw new StackException(this.CreateStackTrace(), "Unable to pop stack, empty");
            }

            return obj;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public T PopStack<T>() where T : IValue
        {
            var obj = this.PopStack();
            if (obj.GetType() == typeof(T))
            {
                return (T)obj;
            }

            throw new StackException(this.CreateStackTrace(), $"Unable to pop stack, type cast error: wanted {typeof(T).FullName} and got {obj.GetType().FullName}");
        }

        public delegate T CastValueDelegate<T>(IValue input);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public T PopStack<T>(CastValueDelegate<T> castFunc) where T : IValue
        {
            var obj = this.PopStack();
            try
            {
                return castFunc(obj);
            }
            catch (Exception exp)
            {
                throw new StackException(this.CreateStackTrace(), $"Unable to pop stack, type cast error: {exp.Message}");
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public IValue PeekStack()
        {
            if (!this.stack.TryPeek(out var obj))
            {
                throw new StackException(this.CreateStackTrace(), "Unable to peek stack, empty");
            }

            return obj;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public T PeekStack<T>() where T : IValue
        {
            var obj = this.PeekStack();
            if (obj.GetType() == typeof(T))
            {
                return (T)obj;
            }

            throw new StackException(this.CreateStackTrace(), "Unable to peek stack, type cast error");
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void PushStack(IValue value)
        {
            if (!this.stack.TryPush(value))
            {
                throw new StackException(this.CreateStackTrace(), "Unable to push stack, stack is full");
            }
        }
        #endregion

        #region Debugging Methods
        public IReadOnlyList<string> CreateStackTrace()
        {
            var result = new List<string>();

            result.Add(DebugScopeLine(this.CurrentCode, this.lineCounter - 1));
            for (var i = this.stackTrace.Index - 1; i >= 0; i--)
            {
                var stackFrame = this.stackTrace.Data[i];
                result.Add(DebugScopeLine(stackFrame.Function, stackFrame.LineCounter));
            }

            return result;
        }

        public IEnumerable<string> PrintStackDebug()
        {
            yield return ($"Stack size: {this.stack.StackSize}");
            for (var i = 0; i < this.stack.StackSize; i++)
            {
                var item = this.stack.Data[i];
                yield return ($"- {item.ToString()}");
            }
        }

        private static string DebugScopeLine(Function function, int line)
        {
            if (line >= function.Code.Count)
            {
                return $"[{function.Name}:{line - 1}: end of code";
            }
            if (line < 0)
            {
                return $"[{function.Name}:{line - 1}: before start of code";
            }

            var codeLine = function.Code[line];
            var codeLineInput = codeLine.Input != null ? codeLine.Input.ToString() : "<empty>";
            return $"[{function.Name}]:{line - 1}:{codeLine.Operator}: [{codeLineInput}]";
        }
        #endregion

        #endregion
    }
}